### PR TITLE
[8.0][IMP][mass_mailing_partner] Link mail statistics to partner

### PR DESCRIPTION
**Pull request by antespi**
- Mass mailing stats smart button in partner form.
- Filter and group by partner in mail statistics tree view
- Refactor for OCA guidelines compliance 

cc @Tecnativa

### CODE REVIEW

File: mass_mailing_partner/README.rst [hunk 1]
```diff
@@ -85,4 +96,4 @@ OCA, or the Odoo Community Association, is a nonprofit organization whose
 mission is to support the collaborative development of Odoo features and
 promote its widespread use.
 
-To contribute to this module, please visit http://odoo-community.org.
\ No newline at end of file
+To contribute to this module, please visit http://odoo-community.org.
```
Comment (pedrobaeza) [inline mass_mailing_partner/README.rst]: https

File: mass_mailing_partner/models/mail_mail_statistics.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        partner_id = False
+        obj = self.env[model].browse(res_id)
+        if obj.exists():
+            if model == 'res.partner':
+                partner_id = res_id
+            elif 'partner_id' in obj._fields:
+                partner_id = obj.partner_id.id
+        return partner_id
+
+    @api.multi
```
Comment (pedrobaeza) [inline mass_mailing_partner/models/mail_mail_statistics.py]: You can put here `@api.one`. Odoo has decided to not deprecate this decorator.

File: mass_mailing_partner/models/mail_mass_mailing_contact.py [hunk 3]
```diff
@@ -15,12 +19,13 @@ class MailMassMailingContact(models.Model):
...
     ]
 
-    @api.one
+    @api.multi
     @api.onchange('partner_id')
     def _onchange_partner(self):
-        if self.partner_id:
-            self.name = self.partner_id.name
-            self.email = self.partner_id.email
+        for contact in self:
```
Comment (pedrobaeza) [inline mass_mailing_partner/models/mail_mass_mailing_contact.py]: onchange must be executed on one record, so no need to support multi-record

File: mass_mailing_partner/models/res_partner.py [hunk 4]
```diff
@@ -9,24 +13,39 @@ class ResPartner(models.Model):
...
+        compute='_compute_mass_mailing_contacts_count', store=True)
+    mass_mailing_stats = fields.One2many(
+        string="Mass mailing stats",
+        comodel_name='mail.mail.statistics', inverse_name='partner_id')
+    mass_mailing_stats_count = fields.Integer(
+        string='Mass mailing stats number',
+        compute='_compute_mass_mailing_stats_count', store=True)
 
-    @api.one
+    @api.multi
```
Comment (pedrobaeza) [inline mass_mailing_partner/models/res_partner.py]: Leave it with @api.one

File: mass_mailing_partner/models/mail_mail_statistics.py (hunk 2 above)
Comment (antespi) [inline mass_mailing_partner/models/mail_mail_statistics.py]: Here I prefer to use api.multi and iterate over self

File: mass_mailing_partner/hooks.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    with api.Environment.manage():
+        env = api.Environment(cr, SUPERUSER_ID, {})
+        # ACTION 1: Match existing contacts
+        contact_model = env['mail.mass_mailing.contact']
+        partner_model = env['res.partner']
+        contacts = contact_model.search([('email', '!=', False)])
+        _logger.info('Trying to match %d contacts to partner by email',
+                     len(contacts))
+        for contact in contacts:
+            partners = partner_model.search([('email', '=ilike',
```
Comment (pedrobaeza) [inline mass_mailing_partner/hooks.py]: Better to make the search directly with the parameter `limit=1`

**Comment by pedrobaeza** [comment]
You should test the post_init_hook.

File: mass_mailing_partner/tests/test_res_partner.py [hunk 6]
```diff
@@ -15,7 +15,7 @@ def test_count_mass_mailing_contacts(self):
                                      'list_id': self.mailing_list.id})
         self.create_mailing_contact({'email': 'partner@test.com',
                                      'list_id': self.mailing_list2.id})
-        self.partner._count_mass_mailing_contacts()
+        self.partner._compute_mass_mailing_contacts_count()
```
Comment (pedrobaeza) [inline mass_mailing_partner/tests/test_res_partner.py]: Why do you need to call the compute method?

**Comment by pedrobaeza** [comment]
For completing the test coverage, do these 2 things in mail_mass_mailing_contact.py:
- Test _onchange_partner method (https://coveralls.io/builds/6992339/source?filename=mass_mailing_partner%2Fmodels%2Fmail_mass_mailing_contact.py#L25)
- Put `# pragma: no cover` on the line https://coveralls.io/builds/6992339/source?filename=mass_mailing_partner%2Fmodels%2Fmail_mass_mailing_contact.py#L57

File: mass_mailing_partner/hooks.py [hunk 7]
```diff
@@ -21,10 +21,11 @@ def post_init_hook(cr, registry):
         _logger.info('Trying to match %d contacts to partner by email',
                      len(contacts))
         for contact in contacts:
-            partners = partner_model.search([('email', '=ilike',
-                                              contact.email)])
+            partners = partner_model.search([
```
Comment (pedrobaeza) [inline mass_mailing_partner/hooks.py:24]: Rename this to `partner` now that the result is only one

**Comment by rafaelbn** [comment]
Please review coveralls
Functionally :+1:

**Comment by yajo** [comment]
It seems you only need tests for this to get :green_heart:

File: mass_mailing_partner/models/mail_mail_statistics.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    def partner_link(self):
+        for stat in self:
+            if stat.model and stat.res_id:
```
Comment (osvalr) [inline mass_mailing_partner/models/mail_mail_statistics.py]: Using `filtered()` on `self` will help to loop for necessary records, don't you think? removing this `if`

Comment (antespi) [inline mass_mailing_partner/models/mail_mail_statistics.py]: Done, thanks :smile:

**Comment by antespi** [comment]
Rebased and squashed for merging

**Comment by antespi** [comment]
All green :four_leaf_clover:

**Comment by yajo** [comment]
:+1: 

Note @osvalr, those reactions do not count as votes for the LGTM status.

**Comment by antespi** [comment]
100% coveralls of this addon:
image

File: mass_mailing_partner/tests/test_res_partner.py (hunk 6 above)
Comment (pedrobaeza) [inline mass_mailing_partner/tests/test_res_partner.py]: This question is not answered

Comment (antespi) [inline mass_mailing_partner/tests/test_res_partner.py]: You're right, this call is not necessary. I've removed it

**Comment by pedrobaeza** [comment]
:+1:

**Pull request merged**